File: src/web/endpoint.ts
import { createHash, randomUUID } from "node:crypto";
import {
  mkdirSync,
  readFileSync,
  renameSync,
  rmSync,
  statSync,
  writeFileSync,
  type Stats,
} from "node:fs";
import { readFile, rename, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { pluginStateDir } from "../config";
import { RUN_UUID_PATTERN } from "../history/types";
import { canonicalRepoRoot } from "../history/store";
import {
  assertCredentialStoreSafe,
  assertPrivateCredentialFile,
  assertPrivateCredentialFileSync,
} from "../fs-private";
import { startWebServer, type WebServer } from "./server";

type EndpointRecord = {
  repoRoot: string;
  url: string;
  /**
   * Identity of the build serving this endpoint, absent when the owner had none to claim. An
   * adopting client compares it against its own, so a workbench never serves a build its caller
   * did not ask for. This is what keeps the invariant, not the owner noticing its own code change.
   */
  build?: string;
};

export type WorkbenchHandle = {
  url: string;
  owned: boolean;
  stop: () => void;
};

type EndpointLockHold = {
  base: string;
  token: string;
};

export type EnsureWorkbenchDeps = {
  start?: (opts: { repoRoot: string; port?: number }) => Promise<WebServer>;
  fetch?: typeof globalThis.fetch;
  sleep?: (ms: number) => Promise<void>;
  stateDir?: string;
  writeRecord?: (record: EndpointRecord, stateDir: string) => Promise<void>;
  now?: () => number;
  staleLockMs?: number;
  lockAttempts?: number;
  lockWaitMs?: number;
};

type AcquireLockHooks = {
  /** Invoked after a stale/dangling decision, before the atomic steal. Tests use this as a barrier. */
  beforeSteal?: (info: { kind: "owned" | "dangling" | "legacy"; token?: string }) => void;
};

const LOCK_ATTEMPTS = 50;
const LOCK_WAIT_MS = 100;
const STALE_LOCK_MS = 10_000;

function endpointKey(repoRoot: string): string {
  return createHash("sha256").update(repoRoot).digest("hex");
}

function endpointsDir(stateDir: string): string {
  return join(stateDir, "web-endpoints");
}

export function endpointRecordPath(repoRoot: string, stateDir: string): string {
  return join(endpointsDir(stateDir), `${endpointKey(repoRoot)}.json`);
}

export function endpointLockPath(repoRoot: string, stateDir: string): string {
  return join(endpointsDir(stateDir), `${endpointKey(repoRoot)}.lock`);
}

function ownedLockPath(base: string, token: string): string {
  return `${base}.${token}`;
}

async function ensurePrivateDir(stateDir: string): Promise<void> {
  await assertCredentialStoreSafe(stateDir);
  await assertCredentialStoreSafe(endpointsDir(stateDir));
}

async function writePrivateFile(path: string, body: string): Promise<void> {
  const tmp = `${path}.${process.pid}.${Date.now()}.tmp`;
  await writeFile(tmp, body, { mode: 0o600 });
  try {
    await assertPrivateCredentialFile(tmp);
    await rename(tmp, path);
    await assertPrivateCredentialFile(path);
  } catch (error) {
    await rm(tmp, { force: true }).catch(() => undefined);
    await rm(path, { force: true }).catch(() => undefined);
    throw error;
  }
}

export async function readEndpointRecord(
  repoRoot: string,
  stateDir: string = pluginStateDir(),
): Promise<EndpointRecord | undefined> {
  const path = endpointRecordPath(repoRoot, stateDir);
  try {
    const raw = await readFile(path, "utf8");
    const parsed = JSON.parse(raw) as unknown;
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) return undefined;
    const row = parsed as Record<string, unknown>;
    if (typeof row.repoRoot !== "string" || typeof row.url !== "string") return undefined;
    if (!row.repoRoot || !row.url) return undefined;
    const build = typeof row.build === "string" && row.build ? row.build : undefined;
    return { repoRoot: row.repoRoot, url: row.url, ...(build ? { build } : {}) };
  } catch {
    return undefined;
  }
}

export async function writeEndpointRecord(
  record: EndpointRecord,
  stateDir: string = pluginStateDir(),
): Promise<void> {
  await ensurePrivateDir(stateDir);
  await writePrivateFile(endpointRecordPath(record.repoRoot, stateDir), JSON.stringify(record));
}

/** Drop the record only when it still names this owner's URL. Caller must hold the endpoint lock. */
function removeEndpointRecordIfUrl(repoRoot: string, stateDir: string, url: string): void {
  const path = endpointRecordPath(repoRoot, stateDir);
  try {
    const raw = readFileSync(path, "utf8");
    const parsed = JSON.parse(raw) as unknown;
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) return;
    if ((parsed as { url?: unknown }).url !== url) return;
    rmSync(path, { force: true });
  } catch {
    // missing or unreadable — nothing this owner should clear
  }
}

export async function probeEndpoint(
  url: string,
  expectedRepoRoot: string,
  fetchImpl: typeof globalThis.fetch = fetch,
): Promise<boolean> {
  try {
    const parsed = new URL(url);
    const token = parsed.searchParams.get("token");
    if (!token) return false;
    const res = await fetchImpl(`${parsed.protocol}//${parsed.host}/api/state`, {
      headers: { "x-hwf-token": token },
      signal: AbortSignal.timeout(2000),
    });
    if (!res.ok) return false;
    const data = (await res.json()) as { canonicalRepoRoot?: unknown };
    return data.canonicalRepoRoot === expectedRepoRoot;
  } catch {
    return false;
  }
}

function isStale(st: Stats, now: () => number, staleLockMs: number): boolean {
  return now() - st.mtimeMs >= staleLockMs;
}

function clearClaimIfToken(base: string, expectedToken: string): void {
  const trash = `${base}.reclaim.${randomUUID()}`;
  try {
    renameSync(base, trash);
  } catch {
    return;
  }
  try {
    if (readFileSync(trash, "utf8").trim() !== expectedToken) {
      try {
        renameSync(trash, base);
      } catch {
        rmSync(trash, { force: true });
      }
      return;
    }
  } catch {
    // unreadable stolen claim
  }
  rmSync(trash, { force: true });
}

/**
 * Atomically steal a stale claim. Contenders race on renaming the unique owned dir
 * (or the claim file when dangling); only one wins, so a successor claim cannot be
 * deleted by a loser still acting on the old token.
 */
function reclaimStaleClaimSync(
  base: string,
  now: () => number,
  staleLockMs: number,
  hooks?: AcquireLockHooks,
): boolean {
  let st: Stats;
  try {
    st = statSync(base);
  } catch {
    return false;
  }

  if (st.isDirectory()) {
    if (!isStale(st, now, staleLockMs)) return false;
    hooks?.beforeSteal?.({ kind: "legacy" });
    const trash = `${base}.reclaim.${randomUUID()}`;
    try {
      renameSync(base, trash);
    } catch {
      return false;
    }
    if (!isStale(statSync(trash), now, staleLockMs)) {
      try {
        renameSync(trash, base);
      } catch {
        rmSync(trash, { recursive: true, force: true });
      }
      return false;
    }
    rmSync(trash, { recursive: true, force: true });
    return true;
  }

  let oldToken: string;
  try {
    oldToken = readFileSync(base, "utf8").trim();
  } catch {
    return false;
  }
  if (!oldToken) return false;

  const owned = ownedLockPath(base, oldToken);
  let ownedSt: Stats | undefined;
  try {
    ownedSt = statSync(owned);
  } catch {
    ownedSt = undefined;
  }

  if (ownedSt && !isStale(ownedSt, now, staleLockMs)) return false;

  if (!ownedSt) {
    hooks?.beforeSteal?.({ kind: "dangling", token: oldToken });
    clearClaimIfToken(base, oldToken);
    return !existsClaim(base);
  }

  hooks?.beforeSteal?.({ kind: "owned", token: oldToken });
  const trashOwned = `${owned}.reclaim.${randomUUID()}`;
  try {
    renameSync(owned, trashOwned);
  } catch {
    return false;
  }

  if (!isStale(statSync(trashOwned), now, staleLockMs)) {
    try {
      renameSync(trashOwned, owned);
    } catch {
      rmSync(trashOwned, { recursive: true, force: true });
    }
    return false;
  }

  clearClaimIfToken(base, oldToken);
  rmSync(trashOwned, { recursive: true, force: true });
  return true;
}

function existsClaim(base: string): boolean {
  try {
    statSync(base);
    return true;
  } catch {
    return false;
  }
}

/**
 * Claim file at `base` stores token; unique dir at `base.<token>` is the only path release deletes.
 * Stale reclaim steals via rename of that unique owned dir so only one contender can clear the claim.
 */
export function acquireEndpointLockSync(
  base: string,
  now: () => number = Date.now,
  staleLockMs: number = STALE_LOCK_MS,
  hooks?: AcquireLockHooks,
): EndpointLockHold | undefined {
  const token = randomUUID();
  const mine = ownedLockPath(base, token);
  mkdirSync(mine);

  const tryClaim = (): boolean => {
    try {
      writeFileSync(base, token, { flag: "wx", mode: 0o600 });
      assertPrivateCredentialFileSync(base);
      return true;
    } catch (error) {
      const code = (error as NodeJS.ErrnoException).code;
      if (code !== "EEXIST") {
        try {
          rmSync(base, { force: true });
        } catch {
          /* best-effort */
        }
        throw error;
      }
      return false;
    }
  };

  if (tryClaim()) return { base, token };
  reclaimStaleClaimSync(base, now, staleLockMs, hooks);
  if (tryClaim()) return { base, token };

  rmSync(mine, { recursive: true, force: true });
  return undefined;
}

/** Deletes only this hold's unique owned directory — never the shared claim path. */
export function releaseEndpointLockSync(hold: EndpointLockHold): void {
  rmSync(ownedLockPath(hold.base, hold.token), { recursive: true, force: true });
}

function clearOwnedRecordUnderLock(
  repoRoot: string,
  stateDir: string,
  url: string,
  now: () => number,
  staleLockMs: number,
): void {
  const hold = acquireEndpointLockSync(endpointLockPath(repoRoot, stateDir), now, staleLockMs);
  if (!hold) return;
  try {
    removeEndpointRecordIfUrl(repoRoot, stateDir, url);
  } finally {
    releaseEndpointLockSync(hold);
  }
}

/**
 * Read-only liveness check — never deletes records. A record whose build differs from the caller's
 * is not adoptable however healthy it is: adopting it would serve code the caller did not ask for.
 */
async function probeLiveRecord(
  repoRoot: string,
  stateDir: string,
  fetchImpl: typeof globalThis.fetch,
  build: string | undefined,
): Promise<EndpointRecord | undefined> {
  const record = await readEndpointRecord(repoRoot, stateDir);
  if (!record) return undefined;
  if (record.repoRoot !== repoRoot) return undefined;
  if (record.build !== build) return undefined;
  if (!(await probeEndpoint(record.url, repoRoot, fetchImpl))) return undefined;
  return record;
}

/** Caller must hold the endpoint lock. Removes only a still-unusable record. */
async function discardUnusableRecord(
  repoRoot: string,
  stateDir: string,
  fetchImpl: typeof globalThis.fetch,
): Promise<void> {
  const record = await readEndpointRecord(repoRoot, stateDir);
  if (!record) return;
  if (record.repoRoot !== repoRoot) {
    await rm(endpointRecordPath(repoRoot, stateDir), { force: true });
    return;
  }
  if (await probeEndpoint(record.url, repoRoot, fetchImpl)) return;
  removeEndpointRecordIfUrl(repoRoot, stateDir, record.url);
}

/** An explicit --port is an instruction, so a live endpoint on another port is not a match. */
function servesPort(url: string, port: number | undefined): boolean {
  if (port === undefined) return true;
  try {
    return new URL(url).port === String(port);
  } catch {
    return false;
  }
}

export async function openWorkbench(
  opts: { repoRoot: string; port?: number; build?: string },
  deps: EnsureWorkbenchDeps = {},
): Promise<WorkbenchHandle> {
  const repoRoot = await canonicalRepoRoot(opts.repoRoot);
  const stateDir = deps.stateDir ?? pluginStateDir();
  const fetchImpl = deps.fetch ?? fetch;
  const sleep = deps.sleep ?? ((ms) => new Promise((r) => setTimeout(r, ms)));
  const start = deps.start ?? startWebServer;
  const writeRecord = deps.writeRecord ?? writeEndpointRecord;
  const now = deps.now ?? Date.now;
  const staleLockMs = deps.staleLockMs ?? STALE_LOCK_MS;
  const lockAttempts = deps.lockAttempts ?? LOCK_ATTEMPTS;
  const lockWaitMs = deps.lockWaitMs ?? LOCK_WAIT_MS;
  const lockBase = endpointLockPath(repoRoot, stateDir);

  await ensurePrivateDir(stateDir);

  for (let attempt = 0; attempt < lockAttempts; attempt++) {
    const existing = await probeLiveRecord(repoRoot, stateDir, fetchImpl, opts.build);
    if (existing && servesPort(existing.url, opts.port)) {
      return { url: existing.url, owned: false, stop: () => undefined };
    }

    const hold = acquireEndpointLockSync(lockBase, now, staleLockMs);
    if (!hold) {
      await sleep(lockWaitMs);
      continue;
    }

    try {
      const again = await probeLiveRecord(repoRoot, stateDir, fetchImpl, opts.build);
      if (again && servesPort(again.url, opts.port)) {
        return { url: again.url, owned: false, stop: () => undefined };
      }

      if (!again) await discardUnusableRecord(repoRoot, stateDir, fetchImpl);

      const server = await start({ repoRoot, port: opts.port });
      const record: EndpointRecord = {
        repoRoot,
        url: server.url,
        ...(opts.build ? { build: opts.build } : {}),
      };
      try {
        await writeRecord(record, stateDir);
      } catch (error) {
        server.stop();
        throw error;
      }

      let stopped = false;
      const ownedUrl = server.url;
      const stop = () => {
        if (stopped) return;
        stopped = true;
        server.stop();
        clearOwnedRecordUnderLock(repoRoot, stateDir, ownedUrl, now, staleLockMs);
      };
      return { url: server.url, owned: true, stop };
    } finally {
      releaseEndpointLockSync(hold);
    }
  }

  throw new Error("timed out waiting for repository workbench endpoint");
}

const SCOPED_ROUTE_RE = /^(w|share)=(repo|global):([a-z0-9][a-z0-9-_]*)$/;
const RUN_ROUTE_RE = new RegExp(`^run=(${RUN_UUID_PATTERN})$`, "i");

export type WebRoute =
  | { kind: "w" | "share"; scope: "repo" | "global"; name: string; hash: string }
  | { kind: "import"; hash: "import" }
  | { kind: "new"; hash: "new" }
  | { kind: "run"; id: string; hash: string };

export function parseWebRoute(raw: string): WebRoute | undefined {
  if (raw === "import") return { kind: "import", hash: "import" };
  if (raw === "new") return { kind: "new", hash: "new" };
  const run = RUN_ROUTE_RE.exec(raw);
  if (run) {
    const id = run[1]!.toLowerCase();
    return { kind: "run", id, hash: `run=${id}` };
  }
  if (raw.startsWith("run=")) return undefined;
  const m = SCOPED_ROUTE_RE.exec(raw);
  if (!m) return undefined;
  const kind = m[1] as "w" | "share";
  const scope = m[2] as "repo" | "global";
  const name = m[3]!;
  return { kind, scope, name, hash: `${kind}=${scope}:${name}` };
}

export function appendRouteHash(url: string, route: WebRoute | undefined): string {
  if (!route) return url;
  return `${url}#${route.hash}`;
}

export function runWorkbenchRoute(id: string): string {
  return `run=${id.toLowerCase()}`;
}

File: internal/workbench/endpoint.go
package workbench

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/aorumbayev/herdr-workflows/internal/config"
	"github.com/aorumbayev/herdr-workflows/internal/credentials"
	"github.com/aorumbayev/herdr-workflows/internal/history"
)

const (
	lockAttemptsDefault = 50
	lockWaitDefault     = 100 * time.Millisecond
	staleLockDefault    = 10 * time.Second
	probeTimeoutDefault = 2 * time.Second
)

// EndpointRecord names one published workbench endpoint.
type EndpointRecord struct {
	RepoRoot string `json:"repoRoot"`
	URL      string `json:"url"`
	Build    string `json:"build,omitempty"`
}

// WorkbenchHandle is an adopted or owned workbench endpoint.
type WorkbenchHandle struct {
	URL   string
	Owned bool
	Stop  func()
}

// HTTPDoer performs HTTP requests. nil uses http.DefaultClient.
type HTTPDoer interface {
	Do(*http.Request) (*http.Response, error)
}

// EnsureWorkbenchDeps injects seams for tests and callers.
type EnsureWorkbenchDeps struct {
	Start        func(Options) (*Server, error)
	HTTPClient   HTTPDoer
	Sleep        func(time.Duration)
	StateDir     string
	WriteRecord  func(EndpointRecord, string) error
	Now          func() time.Time
	StaleLock    time.Duration
	LockAttempts int
	LockWait     time.Duration
}

// OpenWorkbenchOptions selects a repository workbench endpoint.
type OpenWorkbenchOptions struct {
	RepoRoot string
	Port     int
	Build    string
}

func endpointKey(repoRoot string) string {
	sum := sha256.Sum256([]byte(repoRoot))
	return hex.EncodeToString(sum[:])
}

func endpointsDir(stateDir string) string {
	return filepath.Join(stateDir, "web-endpoints")
}

// EndpointRecordPath is the on-disk endpoint record for one repository.
func EndpointRecordPath(repoRoot, stateDir string) string {
	return filepath.Join(endpointsDir(stateDir), endpointKey(repoRoot)+".json")
}

// EndpointLockPath is the on-disk lock file for one repository.
func EndpointLockPath(repoRoot, stateDir string) string {
	return filepath.Join(endpointsDir(stateDir), endpointKey(repoRoot)+".lock")
}

func resolveStateDir(deps EnsureWorkbenchDeps) (string, error) {
	if deps.StateDir != "" {
		return deps.StateDir, nil
	}
	return config.PluginStateDir(os.Getenv)
}

func ensurePrivateDir(stateDir string) error {
	if err := credentials.AssertCredentialStoreSafe(stateDir, nil); err != nil {
		return err
	}
	return credentials.AssertCredentialStoreSafe(endpointsDir(stateDir), nil)
}

func writePrivateFile(path, body string) error {
	tmp := fmt.Sprintf("%s.%d.%d.tmp", path, os.Getpid(), time.Now().UnixNano())
	if err := os.WriteFile(tmp, []byte(body), 0o600); err != nil {
		return err
	}
	if err := credentials.AssertPrivateCredentialFile(tmp, nil); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	if err := os.Rename(tmp, path); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	return credentials.AssertPrivateCredentialFile(path, nil)
}

// ReadEndpointRecord loads the published endpoint for repoRoot, if any.
func ReadEndpointRecord(repoRoot, stateDir string) (*EndpointRecord, error) {
	path := EndpointRecordPath(repoRoot, stateDir)
	raw, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, nil
	}
	var parsed map[string]json.RawMessage
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, nil
	}
	var root, recordURL, build string
	if err := json.Unmarshal(parsed["repoRoot"], &root); err != nil || root == "" {
		return nil, nil
	}
	if err := json.Unmarshal(parsed["url"], &recordURL); err != nil || recordURL == "" {
		return nil, nil
	}
	if rawBuild, ok := parsed["build"]; ok {
		_ = json.Unmarshal(rawBuild, &build)
		if build == "" {
			build = ""
		}
	}
	rec := &EndpointRecord{RepoRoot: root, URL: recordURL}
	if build != "" {
		rec.Build = build
	}
	return rec, nil
}

// WriteEndpointRecord publishes an endpoint record for repoRoot.
func WriteEndpointRecord(record EndpointRecord, stateDir string) error {
	if err := ensurePrivateDir(stateDir); err != nil {
		return err
	}
	body, err := json.Marshal(record)
	if err != nil {
		return err
	}
	return writePrivateFile(EndpointRecordPath(record.RepoRoot, stateDir), string(body))
}

func removeEndpointRecordIfURL(repoRoot, stateDir, recordURL string) {
	path := EndpointRecordPath(repoRoot, stateDir)
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}
	var parsed struct {
		URL string `json:"url"`
	}
	if err := json.Unmarshal(raw, &parsed); err != nil || parsed.URL != recordURL {
		return
	}
	_ = os.Remove(path)
}

// ProbeEndpoint checks whether url serves expectedRepoRoot.
func ProbeEndpoint(recordURL, expectedRepoRoot string, client HTTPDoer) bool {
	if client == nil {
		client = http.DefaultClient
	}
	parsed, err := url.Parse(recordURL)
	if err != nil {
		return false
	}
	token := parsed.Query().Get("token")
	if token == "" {
		return false
	}
	stateURL := parsed.Scheme + "://" + parsed.Host + "/api/state"
	ctx, cancel := context.WithTimeout(context.Background(), probeTimeoutDefault)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, stateURL, nil)
	if err != nil {
		return false
	}
	req.Header.Set("X-Hwf-Token", token)
	resp, err := client.Do(req)
	if err != nil {
		return false
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return false
	}
	var data struct {
		CanonicalRepoRoot string `json:"canonicalRepoRoot"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return false
	}
	return data.CanonicalRepoRoot == expectedRepoRoot
}

func probeLiveRecord(repoRoot, stateDir string, client HTTPDoer, build string) (*EndpointRecord, error) {
	record, err := ReadEndpointRecord(repoRoot, stateDir)
	if err != nil || record == nil {
		return nil, err
	}
	if record.RepoRoot != repoRoot {
		return nil, nil
	}
	if record.Build != build {
		return nil, nil
	}
	if !ProbeEndpoint(record.URL, repoRoot, client) {
		return nil, nil
	}
	return record, nil
}

func discardUnusableRecord(repoRoot, stateDir string, client HTTPDoer) error {
	record, err := ReadEndpointRecord(repoRoot, stateDir)
	if err != nil || record == nil {
		return err
	}
	if record.RepoRoot != repoRoot {
		return os.Remove(EndpointRecordPath(repoRoot, stateDir))
	}
	if ProbeEndpoint(record.URL, repoRoot, client) {
		return nil
	}
	removeEndpointRecordIfURL(repoRoot, stateDir, record.URL)
	return nil
}

func servesPort(recordURL string, port int) bool {
	if port == 0 {
		return true
	}
	parsed, err := url.Parse(recordURL)
	if err != nil {
		return false
	}
	return parsed.Port() == fmt.Sprintf("%d", port)
}

func clearOwnedRecordUnderLock(repoRoot, stateDir, recordURL string, now time.Time, staleLock time.Duration) {
	hold := AcquireEndpointLockSync(EndpointLockPath(repoRoot, stateDir), now, staleLock, nil)
	if hold == nil {
		return
	}
	defer ReleaseEndpointLockSync(hold)
	removeEndpointRecordIfURL(repoRoot, stateDir, recordURL)
}

// OpenWorkbench adopts a live endpoint or starts one for repoRoot.
func OpenWorkbench(opts OpenWorkbenchOptions, deps EnsureWorkbenchDeps) (*WorkbenchHandle, error) {
	repoRoot := history.CanonicalRepoRoot(opts.RepoRoot)
	stateDir, err := resolveStateDir(deps)
	if err != nil {
		return nil, err
	}
	client := deps.HTTPClient
	if client == nil {
		client = http.DefaultClient
	}
	sleep := deps.Sleep
	if sleep == nil {
		sleep = time.Sleep
	}
	start := deps.Start
	if start == nil {
		start = StartWebServer
	}
	writeRecord := deps.WriteRecord
	if writeRecord == nil {
		writeRecord = WriteEndpointRecord
	}
	now := deps.Now
	if now == nil {
		now = time.Now
	}
	staleLock := deps.StaleLock
	if staleLock <= 0 {
		staleLock = staleLockDefault
	}
	lockAttempts := deps.LockAttempts
	if lockAttempts <= 0 {
		lockAttempts = lockAttemptsDefault
	}
	lockWait := deps.LockWait
	if lockWait <= 0 {
		lockWait = lockWaitDefault
	}
	lockBase := EndpointLockPath(repoRoot, stateDir)

	if err := ensurePrivateDir(stateDir); err != nil {
		return nil, err
	}

	for attempt := 0; attempt < lockAttempts; attempt++ {
		existing, err := probeLiveRecord(repoRoot, stateDir, client, opts.Build)
		if err != nil {
			return nil, err
		}
		if existing != nil && servesPort(existing.URL, opts.Port) {
			return &WorkbenchHandle{URL: existing.URL, Owned: false, Stop: func() {}}, nil
		}

		hold := AcquireEndpointLockSync(lockBase, now(), staleLock, nil)
		if hold == nil {
			sleep(lockWait)
			continue
		}

		again, err := probeLiveRecord(repoRoot, stateDir, client, opts.Build)
		if err != nil {
			ReleaseEndpointLockSync(hold)
			return nil, err
		}
		if again != nil && servesPort(again.URL, opts.Port) {
			ReleaseEndpointLockSync(hold)
			return &WorkbenchHandle{URL: again.URL, Owned: false, Stop: func() {}}, nil
		}

		if again == nil {
			if err := discardUnusableRecord(repoRoot, stateDir, client); err != nil {
				ReleaseEndpointLockSync(hold)
				return nil, err
			}
		}

		server, err := start(Options{RepoRoot: repoRoot, Port: opts.Port})
		if err != nil {
			ReleaseEndpointLockSync(hold)
			return nil, err
		}

		record := EndpointRecord{RepoRoot: repoRoot, URL: server.URL}
		if opts.Build != "" {
			record.Build = opts.Build
		}
		if err := writeRecord(record, stateDir); err != nil {
			server.Stop()
			ReleaseEndpointLockSync(hold)
			return nil, err
		}

		stopped := false
		ownedURL := server.URL
		stop := func() {
			if stopped {
				return
			}
			stopped = true
			server.Stop()
			clearOwnedRecordUnderLock(repoRoot, stateDir, ownedURL, now(), staleLock)
		}
		handle := &WorkbenchHandle{URL: server.URL, Owned: true, Stop: stop}
		ReleaseEndpointLockSync(hold)
		return handle, nil
	}

	return nil, errors.New("timed out waiting for repository workbench endpoint")
}

// EndpointLockHold tracks one endpoint lock claim.
type EndpointLockHold struct {
	Base  string
	Token string
}

// BeforeStealInfo describes a stale lock about to be reclaimed.
type BeforeStealInfo struct {
	Kind  string
	Token string
}

// AcquireLockHooks optional hooks for lock acquisition tests.
type AcquireLockHooks struct {
	BeforeSteal func(BeforeStealInfo)
}

func epOwnedLockPath(base, token string) string {
	return base + "." + token
}

func epIsStale(mod time.Time, now time.Time, staleLock time.Duration) bool {
	return now.Sub(mod) >= staleLock
}

func epClearClaimIfToken(base, expectedToken string) {
	trash := base + ".reclaim." + epRandomToken()
	if err := os.Rename(base, trash); err != nil {
		return
	}
	data, err := os.ReadFile(trash)
	if err != nil || strings.TrimSpace(string(data)) != expectedToken {
		_ = os.Rename(trash, base)
		if err != nil {
			_ = os.Remove(trash)
		}
		return
	}
	_ = os.Remove(trash)
}

func epExistsClaim(base string) bool {
	_, err := os.Stat(base)
	return err == nil
}

func epReclaimStaleClaimSync(base string, now time.Time, staleLock time.Duration, hooks *AcquireLockHooks) bool {
	st, err := os.Stat(base)
	if err != nil {
		return false
	}
	if st.IsDir() {
		return epReclaimLegacyDir(base, st, now, staleLock, hooks)
	}
	return epReclaimOwnedFile(base, now, staleLock, hooks)
}

func epReclaimLegacyDir(base string, st os.FileInfo, now time.Time, staleLock time.Duration, hooks *AcquireLockHooks) bool {
	if !epIsStale(st.ModTime(), now, staleLock) {
		return false
	}
	if hooks != nil && hooks.BeforeSteal != nil {
		hooks.BeforeSteal(BeforeStealInfo{Kind: "legacy"})
	}
	trash := base + ".reclaim." + epRandomToken()
	if err := os.Rename(base, trash); err != nil {
		return false
	}
	trashSt, err := os.Stat(trash)
	if err != nil || !epIsStale(trashSt.ModTime(), now, staleLock) {
		if renameErr := os.Rename(trash, base); renameErr != nil {
			_ = os.RemoveAll(trash)
		}
		return false
	}
	_ = os.RemoveAll(trash)
	return true
}

func epReclaimOwnedFile(base string, now time.Time, staleLock time.Duration, hooks *AcquireLockHooks) bool {
	oldTokenBytes, err := os.ReadFile(base)
	if err != nil {
		return false
	}
	oldToken := strings.TrimSpace(string(oldTokenBytes))
	if oldToken == "" {
		return false
	}

	owned := epOwnedLockPath(base, oldToken)
	ownedSt, err := os.Stat(owned)
	if err != nil {
		if !os.IsNotExist(err) {
			return false
		}
		if hooks != nil && hooks.BeforeSteal != nil {
			hooks.BeforeSteal(BeforeStealInfo{Kind: "dangling", Token: oldToken})
		}
		epClearClaimIfToken(base, oldToken)
		return !epExistsClaim(base)
	}

	if !epIsStale(ownedSt.ModTime(), now, staleLock) {
		return false
	}

	if hooks != nil && hooks.BeforeSteal != nil {
		hooks.BeforeSteal(BeforeStealInfo{Kind: "owned", Token: oldToken})
	}
	trashOwned := owned + ".reclaim." + epRandomToken()
	if err := os.Rename(owned, trashOwned); err != nil {
		return false
	}
	trashOwnedSt, err := os.Stat(trashOwned)
	if err != nil || !epIsStale(trashOwnedSt.ModTime(), now, staleLock) {
		_ = os.Rename(trashOwned, owned)
		if err != nil {
			_ = os.RemoveAll(trashOwned)
		}
		return false
	}

	epClearClaimIfToken(base, oldToken)
	_ = os.RemoveAll(trashOwned)
	return true
}

// AcquireEndpointLockSync claims the endpoint lock at base.
func AcquireEndpointLockSync(base string, now time.Time, staleLock time.Duration, hooks *AcquireLockHooks) *EndpointLockHold {
	if staleLock <= 0 {
		staleLock = staleLockDefault
	}
	token := epRandomToken()
	mine := epOwnedLockPath(base, token)
	if err := os.Mkdir(mine, 0o700); err != nil {
		return nil
	}

	tryClaim := func() (bool, error) {
		f, err := os.OpenFile(base, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
		if err != nil {
			if errors.Is(err, os.ErrExist) {
				return false, nil
			}
			_ = os.Remove(base)
			return false, err
		}
		defer func() { _ = f.Close() }()
		if _, err := io.WriteString(f, token); err != nil {
			_ = os.Remove(base)
			return false, err
		}
		if err := credentials.AssertPrivateCredentialFile(base, nil); err != nil {
			_ = os.Remove(base)
			return false, err
		}
		return true, nil
	}

	claim, err := tryClaim()
	if err != nil {
		_ = os.RemoveAll(mine)
		return nil
	}
	if claim {
		return &EndpointLockHold{Base: base, Token: token}
	}
	epReclaimStaleClaimSync(base, now, staleLock, hooks)
	claim, err = tryClaim()
	if err != nil {
		_ = os.RemoveAll(mine)
		return nil
	}
	if claim {
		return &EndpointLockHold{Base: base, Token: token}
	}
	_ = os.RemoveAll(mine)
	return nil
}

// ReleaseEndpointLockSync releases one endpoint lock hold.
func ReleaseEndpointLockSync(hold *EndpointLockHold) {
	if hold == nil {
		return
	}
	_ = os.RemoveAll(epOwnedLockPath(hold.Base, hold.Token))
}

func epRandomToken() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		panic(err)
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	dst := make([]byte, 36)
	hex.Encode(dst[0:8], b[0:4])
	dst[8] = '-'
	hex.Encode(dst[9:13], b[4:6])
	dst[13] = '-'
	hex.Encode(dst[14:18], b[6:8])
	dst[18] = '-'
	hex.Encode(dst[19:23], b[8:10])
	dst[23] = '-'
	hex.Encode(dst[24:36], b[10:16])
	return string(dst)
}
